### PR TITLE
Replace double grass and large fern with placed block

### LINKED ISSUE #3626: Visual bug on placing blocks
Client version:   1.8
Server OS:        Linux
Commit id:        9fa8477b73ce878c631ba875ec10746b4143ba75

### Expected behavior
The block should replace it and not create another one.

### Actual behavior
Only a single block should be placed, however there are 2 of them.

### Steps to reproduce the behavior
Take a dead bush or a double-tall grass and place a block on it. 

#### Notes:
However, it is actually only one block as when you break the additional block, the block over it will also break. If you break the original block it will be a single-hit-dig and there will be no drops.

####Screenshot:
_Before placing:_

_After placing:_

### PR DESCRIPTION
**Pull request by Seadragon91**
Fixes #3626.
If a block is placed on a big flower, the upper or lower part should be replaced by air. I am not sure, where I have to do that.

### CODE REVIEW

**Comment by Seadragon91** [comment]
Now the upper or lower block is replaced by air. The block is still visible on the client, before it's replaced by air.

File: src/Blocks/BlockBigFlower.h [hunk 1]
```diff
@@ -18,6 +18,11 @@ class cBlockBigFlowerHandler :
 	{
 	}
 
+	virtual bool DoesIgnoreBuildCollision(cPlayer * a_Player, NIBBLETYPE a_Meta) override
+	{
+		return (a_Meta & E_META_BIG_FLOWER_DOUBLE_TALL_GRASS || a_Meta & E_META_BIG_FLOWER_LARGE_FERN);
```
Comment (madmaxoft) [inline src/Blocks/BlockBigFlower.h]: I'd prefer the explicit longer condition:
`((a_Meta & E_META_BIG_FLOWER_DOUBLE_TALL_GRASS) != 0) || ...`
After all, the whole codebase has the style that we don't use integers as booleans, we always use an explicit bool result.

File: src/Blocks/BlockBigFlower.h [hunk 2]
```diff
@@ -54,7 +59,7 @@ class cBlockBigFlowerHandler :
 		NIBBLETYPE FlowerMeta = Meta & 0x7;
 		if (!a_Player->IsGameModeCreative())
 		{
-			if (((FlowerMeta == 2) || (FlowerMeta == 3)) && (a_Player->GetEquippedItem().m_ItemType == E_ITEM_SHEARS))
+			if (((FlowerMeta == E_META_BIG_FLOWER_DOUBLE_TALL_GRASS) || (FlowerMeta == E_META_BIG_FLOWER_LARGE_FERN)) && (a_Player->GetEquippedItem().m_ItemType == E_ITEM_SHEARS))
```
Comment (madmaxoft) [inline src/Blocks/BlockBigFlower.h]: Split this into multiple lines

File: src/Blocks/BlockBigFlower.h [hunk 3]
```diff
@@ -73,6 +78,11 @@ class cBlockBigFlowerHandler :
 				a_Player->UseEquippedItem();
 			}
 		}
+
+		if (((FlowerMeta == E_META_BIG_FLOWER_DOUBLE_TALL_GRASS) || (FlowerMeta == E_META_BIG_FLOWER_LARGE_FERN)) && (a_Player->GetEquippedItem().m_ItemType != E_ITEM_SHEARS))
```
Comment (madmaxoft) [inline src/Blocks/BlockBigFlower.h]: Split into multiple lines

**Comment by Seadragon91** [comment]
Okay for merge`?

File: src/Blocks/BlockBigFlower.h [hunk 4]
```diff
@@ -54,7 +59,11 @@ class cBlockBigFlowerHandler :
 		NIBBLETYPE FlowerMeta = Meta & 0x7;
 		if (!a_Player->IsGameModeCreative())
 		{
-			if (((FlowerMeta == 2) || (FlowerMeta == 3)) && (a_Player->GetEquippedItem().m_ItemType == E_ITEM_SHEARS))
+			if (
+				((FlowerMeta == E_META_BIG_FLOWER_DOUBLE_TALL_GRASS) ||
+				(FlowerMeta == E_META_BIG_FLOWER_LARGE_FERN)) &&
+				(a_Player->GetEquippedItem().m_ItemType == E_ITEM_SHEARS)
```
Comment (madmaxoft) [inline src/Blocks/BlockBigFlower.h]: This is poorly split, it hides the different levels of the condition operators. Consider instead:
```cpp
if (
	(a_Player->GetEquippedItem().m_ItemType == E_ITEM_SHEARS) &&
	(
			(FlowerMeta == E_META_BIG_FLOWER_DOUBLE_TALL_GRASS) ||
			(FlowerMeta == E_META_BIG_FLOWER_LARGE_FERN)
	)
)
```

File: src/Blocks/BlockBigFlower.h [hunk 5]
```diff
@@ -73,6 +82,15 @@ class cBlockBigFlowerHandler :
 				a_Player->UseEquippedItem();
 			}
 		}
+
+		if (
+			((FlowerMeta == E_META_BIG_FLOWER_DOUBLE_TALL_GRASS) ||
+			(FlowerMeta == E_META_BIG_FLOWER_LARGE_FERN)) &&
+			(a_Player->GetEquippedItem().m_ItemType != E_ITEM_SHEARS)
```
Comment (madmaxoft) [inline src/Blocks/BlockBigFlower.h]: Again, poor split.

**Comment by Seadragon91** [comment]
Did the changes you requested, now okay?

**Pull request merged**